// bigchaindb-driver/dist/node/connection.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('./request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var HEADER_BLACKLIST = ['content-type'];

/**
 * Base connection
 */

var Connection = function () {
    function Connection(path) {
        var headers = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

        _classCallCheck(this, Connection);

        this.path = path;
        this.headers = _extends({}, headers);

        Object.keys(headers).forEach(function (header) {
            if (HEADER_BLACKLIST.includes(header.toLowerCase())) {
                throw new Error('Header ' + header + ' is reserved and cannot be set.');
            }
        });
    }

    _createClass(Connection, [{
        key: 'getApiUrls',
        value: function getApiUrls(endpoint) {
            return this.path + {
                'blocks': 'blocks',
                'blocksDetail': 'blocks/%(blockHeight)s',
                'outputs': 'outputs',
                'transactions': 'transactions',
                'transactionsSync': 'transactions?mode=sync',
                'transactionsCommit': 'transactions?mode=commit',
                'transactionsDetail': 'transactions/%(transactionId)s',
                'assets': 'assets',
                'metadata': 'metadata',
                'votes': 'votes'
            }[endpoint];
        }
    }, {
        key: '_req',
        value: function _req(path) {
            var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

            // NOTE: `options.headers` could be undefined, but that's OK.
            options.headers = _extends({}, options.headers, this.headers);
            return (0, _request2.default)(path, options);
        }

        /**
         * @param blockHeight
         */

    }, {
        key: 'getBlock',
        value: function getBlock(blockHeight) {
            return this._req(this.getApiUrls('blocksDetail'), {
                urlTemplateSpec: {
                    blockHeight: blockHeight
                }
            });
        }

        /**
         * @param transactionId
         */

    }, {
        key: 'getTransaction',
        value: function getTransaction(transactionId) {
            return this._req(this.getApiUrls('transactionsDetail'), {
                urlTemplateSpec: {
                    transactionId: transactionId
                }
            });
        }

        /**
         * @param transactionId
         * @param status
         */

    }, {
        key: 'listBlocks',
        value: function listBlocks(transactionId) {
            return this._req(this.getApiUrls('blocks'), {
                query: {
                    transaction_id: transactionId
                }
            });
        }

        /**
         * @param publicKey
         * @param spent
         */

    }, {
        key: 'listOutputs',
        value: function listOutputs(publicKey, spent) {
            var query = {
                public_key: publicKey
                // NOTE: If `spent` is not defined, it must not be included in the
                // query parameters.
            };if (spent !== undefined) {
                query.spent = spent.toString();
            }
            return this._req(this.getApiUrls('outputs'), {
                query: query
            });
        }

        /**
         * @param assetId
         * @param operation
         */

    }, {
        key: 'listTransactions',
        value: function listTransactions(assetId, operation) {
            return this._req(this.getApiUrls('transactions'), {
                query: {
                    asset_id: assetId,
                    operation: operation
                }
            });
        }

        /**
         * @param blockId
         */

    }, {
        key: 'listVotes',
        value: function listVotes(blockId) {
            return this._req(this.getApiUrls('votes'), {
                query: {
                    block_id: blockId
                }
            });
        }

        /**
         * @param transaction
         */

    }, {
        key: 'postTransaction',
        value: function postTransaction(transaction) {
            return this._req(this.getApiUrls('transactions'), {
                method: 'POST',
                jsonBody: transaction
            });
        }

        /**
         * @param transaction
         */

    }, {
        key: 'postTransactionSync',
        value: function postTransactionSync(transaction) {
            return this._req(this.getApiUrls('transactionsSync'), {
                method: 'POST',
                jsonBody: transaction
            });
        }

        /**
         * @param transaction
         */

    }, {
        key: 'postTransactionCommit',
        value: function postTransactionCommit(transaction) {
            return this._req(this.getApiUrls('transactionsCommit'), {
                method: 'POST',
                jsonBody: transaction
            });
        }

        /**
         * @param search
         */

    }, {
        key: 'searchAssets',
        value: function searchAssets(search) {
            return this._req(this.getApiUrls('assets'), {
                query: {
                    search: search
                }
            });
        }

        /**
         * @param search
         */

    }, {
        key: 'searchMetadata',
        value: function searchMetadata(search) {
            return this._req(this.getApiUrls('metadata'), {
                query: {
                    search: search
                }
            });
        }
    }]);

    return Connection;
}();

exports.default = Connection;
