// bigchaindb-driver/dist/node/transaction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _buffer = require('buffer');

var _jsonStableStringify = require('json-stable-stringify');

var _jsonStableStringify2 = _interopRequireDefault(_jsonStableStringify);

var _clone = require('clone');

var _clone2 = _interopRequireDefault(_clone);

var _bs = require('bs58');

var _bs2 = _interopRequireDefault(_bs);

var _cryptoConditions = require('crypto-conditions');

var _cryptoConditions2 = _interopRequireDefault(_cryptoConditions);

var _ccJsonify = require('./utils/ccJsonify');

var _ccJsonify2 = _interopRequireDefault(_ccJsonify);

var _sha256Hash = require('./sha256Hash');

var _sha256Hash2 = _interopRequireDefault(_sha256Hash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Construct Transactions
 */
var Transaction = function () {
    function Transaction() {
        _classCallCheck(this, Transaction);
    }

    _createClass(Transaction, null, [{
        key: 'serializeTransactionIntoCanonicalString',

        /**
         * Canonically serializes a transaction into a string by sorting the keys
         * @param {Object} (transaction)
         * @return {string} a canonically serialized Transaction
         */
        value: function serializeTransactionIntoCanonicalString(transaction) {
            // BigchainDB signs fulfillments by serializing transactions into a
            // "canonical" format where
            var tx = (0, _clone2.default)(transaction);
            // TODO: set fulfillments to null
            // Sort the keys
            return (0, _jsonStableStringify2.default)(tx, function (a, b) {
                return a.key > b.key ? 1 : -1;
            });
        }
    }, {
        key: 'makeInputTemplate',
        value: function makeInputTemplate() {
            var publicKeys = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
            var fulfills = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
            var fulfillment = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;

            return {
                fulfillment: fulfillment,
                fulfills: fulfills,
                'owners_before': publicKeys
            };
        }
    }, {
        key: 'makeTransactionTemplate',
        value: function makeTransactionTemplate() {
            var txTemplate = {
                'id': null,
                'operation': null,
                'outputs': [],
                'inputs': [],
                'metadata': null,
                'asset': null,
                'version': '2.0'
            };
            return txTemplate;
        }
    }, {
        key: 'makeTransaction',
        value: function makeTransaction(operation, asset) {
            var metadata = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
            var outputs = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];
            var inputs = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : [];

            var tx = Transaction.makeTransactionTemplate();
            tx.operation = operation;
            tx.asset = asset;
            tx.metadata = metadata;
            tx.inputs = inputs;
            tx.outputs = outputs;
            return tx;
        }

        /**
         * Generate a `CREATE` transaction holding the `asset`, `metadata`, and `outputs`, to be signed by
         * the `issuers`.
         * @param {Object} asset Created asset's data
         * @param {Object} metadata Metadata for the Transaction
         * @param {Object[]} outputs Array of Output objects to add to the Transaction.
         *                           Think of these as the recipients of the asset after the transaction.
         *                           For `CREATE` Transactions, this should usually just be a list of
         *                           Outputs wrapping Ed25519 Conditions generated from the issuers' public
         *                           keys (so that the issuers are the recipients of the created asset).
         * @param {...string[]} issuers Public key of one or more issuers to the asset being created by this
         *                              Transaction.
         *                              Note: Each of the private keys corresponding to the given public
         *                              keys MUST be used later (and in the same order) when signing the
         *                              Transaction (`signTransaction()`).
         * @returns {Object} Unsigned transaction -- make sure to call signTransaction() on it before
         *                   sending it off!
         */

    }, {
        key: 'makeCreateTransaction',
        value: function makeCreateTransaction(asset, metadata, outputs) {
            var assetDefinition = {
                'data': asset || null
            };

            for (var _len = arguments.length, issuers = Array(_len > 3 ? _len - 3 : 0), _key = 3; _key < _len; _key++) {
                issuers[_key - 3] = arguments[_key];
            }

            var inputs = issuers.map(function (issuer) {
                return Transaction.makeInputTemplate([issuer]);
            });

            return Transaction.makeTransaction('CREATE', assetDefinition, metadata, outputs, inputs);
        }

        /**
         * Create an Ed25519 Cryptocondition from an Ed25519 public key
         * to put into an Output of a Transaction
         * @param {string} publicKey base58 encoded Ed25519 public key for the recipient of the Transaction
         * @param {boolean} [json=true] If true returns a json object otherwise a crypto-condition type
         * @returns {Object} Ed25519 Condition (that will need to wrapped in an Output)
         */

    }, {
        key: 'makeEd25519Condition',
        value: function makeEd25519Condition(publicKey) {
            var json = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

            var publicKeyBuffer = _buffer.Buffer.from(_bs2.default.decode(publicKey));

            var ed25519Fulfillment = new _cryptoConditions2.default.Ed25519Sha256();
            ed25519Fulfillment.setPublicKey(publicKeyBuffer);

            if (json) {
                return (0, _ccJsonify2.default)(ed25519Fulfillment);
            }

            return ed25519Fulfillment;
        }

        /**
         * Create an Output from a Condition.
         * Note: Assumes the given Condition was generated from a
         * single public key (e.g. a Ed25519 Condition)
         * @param {Object} condition Condition (e.g. a Ed25519 Condition from `makeEd25519Condition()`)
         * @param {string} amount Amount of the output
         * @returns {Object} An Output usable in a Transaction
         */

    }, {
        key: 'makeOutput',
        value: function makeOutput(condition) {
            var amount = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '1';

            if (typeof amount !== 'string') {
                throw new TypeError('`amount` must be of type string');
            }
            var publicKeys = [];
            var getPublicKeys = function getPublicKeys(details) {
                if (details.type === 'ed25519-sha-256') {
                    if (!publicKeys.includes(details.public_key)) {
                        publicKeys.push(details.public_key);
                    }
                } else if (details.type === 'threshold-sha-256') {
                    details.subconditions.map(getPublicKeys);
                }
            };
            getPublicKeys(condition.details);
            return {
                condition: condition,
                'amount': amount,
                'public_keys': publicKeys
            };
        }

        /**
         * Create a Preimage-Sha256 Cryptocondition from a secret to put into an Output of a Transaction
         * @param {string} preimage Preimage to be hashed and wrapped in a crypto-condition
         * @param {boolean} [json=true] If true returns a json object otherwise a crypto-condition type
         * @returns {Object} Preimage-Sha256 Condition (that will need to wrapped in an Output)
         */

    }, {
        key: 'makeSha256Condition',
        value: function makeSha256Condition(preimage) {
            var json = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

            var sha256Fulfillment = new _cryptoConditions2.default.PreimageSha256();
            sha256Fulfillment.preimage = _buffer.Buffer.from(preimage);

            if (json) {
                return (0, _ccJsonify2.default)(sha256Fulfillment);
            }
            return sha256Fulfillment;
        }

        /**
         * Create an Sha256 Threshold Cryptocondition from threshold to put into an Output of a Transaction
         * @param {number} threshold
         * @param {Array} [subconditions=[]]
         * @param {boolean} [json=true] If true returns a json object otherwise a crypto-condition type
         * @returns {Object} Sha256 Threshold Condition (that will need to wrapped in an Output)
         */

    }, {
        key: 'makeThresholdCondition',
        value: function makeThresholdCondition(threshold) {
            var subconditions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
            var json = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

            var thresholdCondition = new _cryptoConditions2.default.ThresholdSha256();
            thresholdCondition.threshold = threshold;

            subconditions.forEach(function (subcondition) {
                // TODO: add support for Condition and URIs
                thresholdCondition.addSubfulfillment(subcondition);
            });

            if (json) {
                return (0, _ccJsonify2.default)(thresholdCondition);
            }

            return thresholdCondition;
        }

        /**
         * Generate a `TRANSFER` transaction holding the `asset`, `metadata`, and `outputs`, that fulfills
         * the `fulfilledOutputs` of `unspentTransaction`.
         * @param {Object} unspentTransaction Previous Transaction you have control over (i.e. can fulfill
         *                                    its Output Condition)
         * @param {Object} metadata Metadata for the Transaction
         * @param {Object[]} outputs Array of Output objects to add to the Transaction.
         *                           Think of these as the recipients of the asset after the transaction.
         *                           For `TRANSFER` Transactions, this should usually just be a list of
         *                           Outputs wrapping Ed25519 Conditions generated from the public keys of
         *                           the recipients.
         * @param {...number} OutputIndices Indices of the Outputs in `unspentTransaction` that this
         *                                     Transaction fulfills.
         *                                     Note that listed public keys listed must be used (and in
         *                                     the same order) to sign the Transaction
         *                                     (`signTransaction()`).
         * @returns {Object} Unsigned transaction -- make sure to call signTransaction() on it before
         *                   sending it off!
         */
        // TODO:
        // - Make `metadata` optional argument

    }, {
        key: 'makeTransferTransaction',
        value: function makeTransferTransaction(unspentOutputs, outputs, metadata) {
            var inputs = unspentOutputs.map(function (unspentOutput) {
                var _tx$outputIndex = { tx: unspentOutput.tx, outputIndex: unspentOutput.output_index },
                    tx = _tx$outputIndex.tx,
                    outputIndex = _tx$outputIndex.outputIndex;

                var fulfilledOutput = tx.outputs[outputIndex];
                var transactionLink = {
                    'output_index': outputIndex,
                    'transaction_id': tx.id
                };

                return Transaction.makeInputTemplate(fulfilledOutput.public_keys, transactionLink);
            });

            var assetLink = {
                'id': unspentOutputs[0].tx.operation === 'CREATE' ? unspentOutputs[0].tx.id : unspentOutputs[0].tx.asset.id
            };
            return Transaction.makeTransaction('TRANSFER', assetLink, metadata, outputs, inputs);
        }

        /**
         * Sign the given `transaction` with the given `privateKey`s, returning a new copy of `transaction`
         * that's been signed.
         * Note: Only generates Ed25519 Fulfillments. Thresholds and other types of Fulfillments are left as
         * an exercise for the user.
         * @param {Object} transaction Transaction to sign. `transaction` is not modified.
         * @param {...string} privateKeys Private keys associated with the issuers of the `transaction`.
         *                                Looped through to iteratively sign any Input Fulfillments found in
         *                                the `transaction`.
         * @returns {Object} The signed version of `transaction`.
         */

    }, {
        key: 'signTransaction',
        value: function signTransaction(transaction) {
            for (var _len2 = arguments.length, privateKeys = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
                privateKeys[_key2 - 1] = arguments[_key2];
            }

            var signedTx = (0, _clone2.default)(transaction);
            signedTx.inputs.forEach(function (input, index) {
                var privateKey = privateKeys[index];
                var privateKeyBuffer = _buffer.Buffer.from(_bs2.default.decode(privateKey));
                var serializedTransaction = Transaction.serializeTransactionIntoCanonicalString(transaction);

                var transactionUniqueFulfillment = input.fulfills ? serializedTransaction.concat(input.fulfills.transaction_id).concat(input.fulfills.output_index) : serializedTransaction;
                var transactionHash = (0, _sha256Hash2.default)(transactionUniqueFulfillment);
                var ed25519Fulfillment = new _cryptoConditions2.default.Ed25519Sha256();
                ed25519Fulfillment.sign(_buffer.Buffer.from(transactionHash, 'hex'), privateKeyBuffer);
                var fulfillmentUri = ed25519Fulfillment.serializeUri();

                input.fulfillment = fulfillmentUri;
            });

            var serializedTransaction = Transaction.serializeTransactionIntoCanonicalString(signedTx);
            signedTx.id = (0, _sha256Hash2.default)(serializedTransaction);
            return signedTx;
        }
    }]);

    return Transaction;
}();

exports.default = Transaction;
